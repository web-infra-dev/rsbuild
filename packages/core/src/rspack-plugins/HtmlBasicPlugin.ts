import type HtmlWebpackPlugin from 'html-webpack-plugin';
import type { Compiler, Compilation } from '@rspack/core';

export type HtmlInfo = {
  title?: string;
  favicon?: string;
  templateContent?: string;
};

export type HtmlBasicPluginOptions = {
  info: Record<string, HtmlInfo>;
  HtmlPlugin: typeof HtmlWebpackPlugin;
};

export const hasTitle = (html?: string): boolean =>
  html ? /<title/i.test(html) && /<\/title/i.test(html) : false;

export class HtmlBasicPlugin {
  readonly name: string;

  readonly options: HtmlBasicPluginOptions;

  readonly HtmlPlugin: typeof HtmlWebpackPlugin;

  constructor(options: HtmlBasicPluginOptions) {
    this.name = 'HtmlBasicPlugin';
    this.options = options;
    this.HtmlPlugin = options.HtmlPlugin;
  }

  apply(compiler: Compiler) {
    const addTitleTag = (
      headTags: HtmlWebpackPlugin.HtmlTagObject[],
      outputName: string,
    ) => {
      const { title } = this.options.info[outputName];
      if (title) {
        headTags.unshift({
          tagName: 'title',
          innerHTML: title,
          attributes: {},
          voidTag: false,
          meta: {},
        });
      }
    };

    const addFavicon = (
      headTags: HtmlWebpackPlugin.HtmlTagObject[],
      outputName: string,
    ) => {
      const { favicon } = this.options.info[outputName];
      if (favicon) {
        headTags.unshift({
          tagName: 'link',
          voidTag: true,
          attributes: {
            rel: 'icon',
            href: favicon,
          },
          meta: {},
        });
      }
    };

    compiler.hooks.compilation.tap(this.name, (compilation: Compilation) => {
      // @ts-expect-error compilation type mismatch
      this.HtmlPlugin.getHooks(compilation).alterAssetTagGroups.tap(
        this.name,
        (data) => {
          const { headTags, outputName } = data;
          const { templateContent } = this.options.info[outputName];

          if (!hasTitle(templateContent)) {
            addTitleTag(headTags, outputName);
          }

          addFavicon(headTags, outputName);
          return data;
        },
      );
    });
  }
}
