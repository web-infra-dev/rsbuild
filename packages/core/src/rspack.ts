import { createRequire } from 'node:module';

/**
 * Currently, Rspack only provides a CJS bundle, so we use require to load it
 * for better startup performance. If we use `import { rspack } from '@rspack/core'`,
 * Node.js will use `cjs-module-lexer` to parse it, which slows down startup by ~30ms.
 * We can remove this module once `@rspack/core` provides an ESM bundle in the future.
 */
const require = createRequire(import.meta.url);
const rspack =
  require('@rspack/core') as (typeof import('@rspack/core'))['rspack'];

export { rspack };
