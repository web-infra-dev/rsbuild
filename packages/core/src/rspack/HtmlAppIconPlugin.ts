import fs from 'fs';
import { posix, basename } from 'path';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import type { Compiler, Compilation } from '@rspack/core';
import WebpackSources from '@rsbuild/shared/webpack-sources';
import {
  withPublicPath,
  getPublicPathFromCompiler,
  COMPILATION_PROCESS_STAGE,
} from '@rsbuild/shared';

type AppIconOptions = {
  distDir: string;
  iconPath: string;
};

export class HtmlAppIconPlugin {
  readonly name: string;

  readonly distDir: string;

  readonly iconPath: string;

  constructor(options: AppIconOptions) {
    this.name = 'HtmlAppIconPlugin';
    this.distDir = options.distDir;
    this.iconPath = options.iconPath;
  }

  apply(compiler: Compiler) {
    if (!fs.existsSync(this.iconPath)) {
      throw new Error(
        `[${this.name}] Can not find the app icon, please check if the '${this.iconPath}' file exists'.`,
      );
    }

    const iconRelativePath = posix.join(this.distDir, basename(this.iconPath));

    // add html asset tags
    compiler.hooks.compilation.tap(this.name, (compilation: Compilation) => {
      // @ts-expect-error compilation type mismatch
      HtmlWebpackPlugin.getHooks(compilation).alterAssetTagGroups.tap(
        this.name,
        (data) => {
          const publicPath = getPublicPathFromCompiler(compiler);

          data.headTags.unshift({
            tagName: 'link',
            voidTag: true,
            attributes: {
              rel: 'apple-touch-icon',
              sizes: '180*180',
              href: withPublicPath(iconRelativePath, publicPath),
            },
            meta: {},
          });

          return data;
        },
      );
    });

    // copy icon to dist
    compiler.hooks.thisCompilation.tap(
      this.name,
      (compilation: Compilation) => {
        compilation.hooks.processAssets.tap(
          {
            name: this.name,
            stage: COMPILATION_PROCESS_STAGE.PROCESS_ASSETS_STAGE_PRE_PROCESS,
          },
          (assets) => {
            const source = fs.readFileSync(this.iconPath);
            assets[iconRelativePath] = new WebpackSources.RawSource(
              source,
              false,
            );
          },
        );
      },
    );
  }
}
