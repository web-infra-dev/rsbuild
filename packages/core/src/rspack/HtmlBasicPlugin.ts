import {
  type HtmlBasicTag,
  type HtmlTag,
  type HtmlTagDescriptor,
  type HtmlTagUtils,
  type ModifyHTMLTagsFn,
  isFunction,
  partition,
} from '@rsbuild/shared';
import type { Compilation, Compiler } from '@rspack/core';
import type HtmlWebpackPlugin from 'html-webpack-plugin';
import type { HtmlTagObject } from 'html-webpack-plugin';
import { ensureAssetPrefix } from '../helpers';
import { getHTMLPlugin } from '../pluginHelper';

export type TagConfig = {
  tags?: HtmlTagDescriptor[];
  hash?: HtmlTag['hash'];
  append?: HtmlTag['append'];
  publicPath?: HtmlTag['publicPath'];
};

/** @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Void_element} */
export const VOID_TAGS = [
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'keygen',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
];

/** @see {@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/head#see_also} */
export const HEAD_TAGS = [
  'title',
  'base',
  'link',
  'style',
  'meta',
  'script',
  'noscript',
  'template',
];

export const FILE_ATTRS = {
  link: 'href',
  script: 'src',
};

export type HtmlInfo = {
  favicon?: string;
  tagConfig?: TagConfig;
  templateContent?: string;
};

export type HtmlBasicPluginOptions = Record<string, HtmlInfo>;

export type AlterAssetTagGroupsData = {
  headTags: HtmlTagObject[];
  bodyTags: HtmlTagObject[];
  outputName: string;
  publicPath: string;
  plugin: HtmlWebpackPlugin;
};

export const hasTitle = (html?: string): boolean =>
  html ? /<title/i.test(html) && /<\/title/i.test(html) : false;

const getTagPriority = (tag: HtmlTag, tagConfig: TagConfig) => {
  const head = tag.head ?? HEAD_TAGS.includes(tag.tag);
  let priority = head ? -2 : 2;

  const append = tag.append ?? tagConfig.append;
  if (typeof append === 'boolean') {
    priority += append ? 1 : -1;
  }

  return priority;
};

/**
 * `HtmlTagObject` -> `HtmlBasicTag`
 */
const formatBasicTag = (tag: HtmlTagObject): HtmlBasicTag => ({
  tag: tag.tagName,
  attrs: tag.attributes,
  children: tag.innerHTML,
});

/**
 * `HtmlBasicTag` -> `HtmlTagObject`
 */
const fromBasicTag = (tag: HtmlBasicTag): HtmlTagObject => ({
  meta: {},
  tagName: tag.tag,
  attributes: tag.attrs ?? {},
  voidTag: VOID_TAGS.includes(tag.tag),
  innerHTML: tag.children,
});

/**
 * `HtmlTagObject[]` -> `HtmlTag[]`
 */
const formatTags = (
  tags: HtmlTagObject[],
  override?: Partial<HtmlTag>,
): HtmlTag[] =>
  tags.map((tag) => ({
    ...formatBasicTag(tag),
    publicPath: false,
    ...override,
  }));

const applyTagConfig = (
  data: AlterAssetTagGroupsData,
  tagConfig: TagConfig,
  compilationHash: string,
  entryName: string,
) => {
  if (!tagConfig.tags?.length) {
    return data;
  }

  const fromInjectTags = (tags: HtmlTag[]) => {
    const ret: HtmlTagObject[] = [];

    for (const tag of tags) {
      // apply publicPath and hash to filename attr.
      const attrs = { ...tag.attrs };
      const filenameTag = FILE_ATTRS[tag.tag as keyof typeof FILE_ATTRS];
      let filename = attrs[filenameTag];

      if (typeof filename === 'string') {
        const optPublicPath = tag.publicPath ?? tagConfig.publicPath;

        if (typeof optPublicPath === 'function') {
          filename = optPublicPath(filename, data.publicPath);
        } else if (typeof optPublicPath === 'string') {
          filename = ensureAssetPrefix(filename, optPublicPath);
        } else if (optPublicPath !== false) {
          filename = ensureAssetPrefix(filename, data.publicPath);
        }

        const optHash = tag.hash ?? tagConfig.hash;

        if (typeof optHash === 'function') {
          if (compilationHash.length) {
            filename = optHash(filename, compilationHash);
          }
        } else if (typeof optHash === 'string') {
          if (optHash.length) {
            filename = `${filename}?${optHash}`;
          }
        } else if (optHash === true) {
          if (compilationHash.length) {
            filename = `${filename}?${compilationHash}`;
          }
        }

        attrs[filenameTag] = filename;
        tag.attrs = attrs;
      }

      ret.push(fromBasicTag(tag));
    }
    return ret;
  };

  let tags = [
    ...formatTags(data.headTags, { head: true }),
    ...formatTags(data.bodyTags, { head: false }),
  ];

  const utils: HtmlTagUtils = {
    hash: compilationHash,
    entryName,
    outputName: data.outputName,
    publicPath: data.publicPath,
  };

  for (const item of tagConfig.tags) {
    if (isFunction(item)) {
      tags = item(tags, utils) || tags;
    } else {
      tags.push(item);
    }

    tags = tags.sort(
      (tag1, tag2) =>
        getTagPriority(tag1, tagConfig) - getTagPriority(tag2, tagConfig),
    );
  }

  const [headTags, bodyTags] = partition(
    tags,
    (tag) => tag.head ?? HEAD_TAGS.includes(tag.tag),
  );
  data.headTags = fromInjectTags(headTags);
  data.bodyTags = fromInjectTags(bodyTags);

  return data;
};

const addTitleTag = (headTags: HtmlTagObject[], title = '') => {
  headTags.unshift({
    tagName: 'title',
    innerHTML: title,
    attributes: {},
    voidTag: false,
    meta: {},
  });
};

const addFavicon = (headTags: HtmlTagObject[], favicon?: string) => {
  if (favicon) {
    headTags.unshift({
      tagName: 'link',
      voidTag: true,
      attributes: {
        rel: 'icon',
        href: favicon,
      },
      meta: {},
    });
  }
};

export class HtmlBasicPlugin {
  readonly name: string;

  readonly options: HtmlBasicPluginOptions;

  readonly modifyTagsFn?: ModifyHTMLTagsFn;

  constructor(
    options: HtmlBasicPluginOptions,
    modifyTagsFn?: ModifyHTMLTagsFn,
  ) {
    this.name = 'HtmlBasicPlugin';
    this.options = options;
    this.modifyTagsFn = modifyTagsFn;
  }

  apply(compiler: Compiler) {
    compiler.hooks.compilation.tap(this.name, (compilation: Compilation) => {
      getHTMLPlugin()
        .getHooks(compilation)
        .alterAssetTagGroups.tapPromise(this.name, async (data) => {
          const entryName = data.plugin.options?.entryName;

          if (!entryName) {
            return data;
          }

          const { headTags, bodyTags } = data;
          const { favicon, tagConfig, templateContent } =
            this.options[entryName];

          if (!hasTitle(templateContent)) {
            addTitleTag(headTags, data.plugin.options?.title);
          }

          addFavicon(headTags, favicon);

          const tags = {
            headTags: headTags.map(formatBasicTag),
            bodyTags: bodyTags.map(formatBasicTag),
          };

          const modified = this.modifyTagsFn
            ? await this.modifyTagsFn(tags, {
                compilation,
                assetPrefix: data.publicPath,
                filename: data.outputName,
              })
            : tags;

          Object.assign(data, {
            headTags: modified.headTags.map(fromBasicTag),
            bodyTags: modified.bodyTags.map(fromBasicTag),
          });

          if (tagConfig) {
            const hash = compilation.hash ?? '';
            applyTagConfig(data, tagConfig, hash, entryName);
          }

          return data;
        });
    });
  }
}
