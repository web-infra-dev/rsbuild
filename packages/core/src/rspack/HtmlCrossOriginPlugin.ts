import { DEFAULT_ASSET_PREFIX, type CrossOrigin } from '@rsbuild/shared';
import type { Compiler, RspackPluginInstance } from '@rspack/core';
import { getHTMLPlugin } from '../provider/htmlPluginUtil';

type CrossOriginOptions = {
  crossOrigin: CrossOrigin;
};

export class HtmlCrossOriginPlugin implements RspackPluginInstance {
  readonly name: string;

  readonly crossOrigin: CrossOrigin;

  constructor(options: CrossOriginOptions) {
    const { crossOrigin } = options;
    this.name = 'HtmlCrossOriginPlugin';
    this.crossOrigin = crossOrigin;
  }

  apply(compiler: Compiler): void {
    if (!this.crossOrigin) {
      return;
    }

    // no need to set crossorigin="anonymous" if asset domain is the same as
    // the page domain, align with webpack's crossOriginLoading logic.
    // see: https://github.com/webpack/webpack/blob/611bded369b5a9ea695f2669b25a6f88b67d7826/lib/runtime/LoadScriptRuntimeModule.js#L100-L111
    if (this.crossOrigin !== 'use-credentials') {
      const { publicPath } = compiler.options.output;
      if (
        !publicPath ||
        publicPath === DEFAULT_ASSET_PREFIX ||
        publicPath === 'auto'
      ) {
        return;
      }
    }

    compiler.hooks.compilation.tap(this.name, (compilation) => {
      getHTMLPlugin()
        // @ts-expect-error compilation type mismatch
        .getHooks(compilation)
        .alterAssetTags.tap(this.name, (alterAssetTags) => {
          const {
            assetTags: { scripts, styles },
          } = alterAssetTags;

          scripts.forEach((script) => {
            script.attributes.crossorigin ??= this.crossOrigin;
          });
          styles.forEach((style) => {
            style.attributes.crossorigin ??= this.crossOrigin;
          });

          return alterAssetTags;
        });
    });
  }
}
