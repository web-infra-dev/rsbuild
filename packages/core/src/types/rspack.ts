import type * as Rspack from '@rspack/core';
import type RspackChain from 'rspack-chain';
import type { TransformHandler } from './plugin';

export type { Rspack, RspackChain };

declare module '@rspack/core' {
  interface Compiler {
    __rsbuildTransformer?: Record<string, TransformHandler>;
  }
}

export interface BundlerPluginInstance {
  [index: string]: any;
  apply: (compiler: any) => void;
}

/** T[] => T */
type GetElementType<T extends any[]> = T extends (infer U)[] ? U : never;

export type RspackRule = GetElementType<
  NonNullable<NonNullable<Rspack.Configuration['module']>['rules']>
>;
