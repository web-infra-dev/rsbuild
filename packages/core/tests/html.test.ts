import { hasTitle } from '@src/rspack/HtmlBasicPlugin';

test('should detect HTML title via "hasTitle" correctly', () => {
  expect(hasTitle()).toEqual(false);
  expect(hasTitle('<title></title>')).toEqual(true);
  expect(hasTitle('<title>hello</title>')).toEqual(true);
  expect(hasTitle('<title  ></title>')).toEqual(true);
  expect(hasTitle('<title></title  >')).toEqual(true);
  expect(hasTitle('<title >hello</title  >')).toEqual(true);
  expect(hasTitle('<html><head></head><body></body></html>')).toEqual(false);
  expect(
    hasTitle(
      '<html><head><title>Page Title</title></head><body></body></html>',
    ),
  ).toEqual(true);
  expect(
    hasTitle(
      '<html>\n<head>\n<title>\nPage Title\n</title>\n</head>\n<body></body>\n</html>',
    ),
  ).toEqual(true);
});
