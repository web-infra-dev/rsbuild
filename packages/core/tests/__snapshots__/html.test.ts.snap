// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-html > should allow to configure html.tags 1`] = `
{
  "entry": {
    "foo": [
      "src/foo.ts",
    ],
    "main": [
      "src/main.ts",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "main",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "main",
        "excludeChunks": [],
        "favicon": false,
        "filename": "main.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "main",
        ],
        "entryName": "main",
        "filename": "main.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "foo",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "foo",
        "excludeChunks": [],
        "favicon": false,
        "filename": "foo.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "foo",
        ],
        "entryName": "foo",
        "filename": "foo.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "foo": {
          "tagConfig": {
            "append": true,
            "hash": false,
            "publicPath": true,
            "tags": [
              {
                "attrs": {
                  "src": "jq.js",
                },
                "tag": "script",
              },
            ],
          },
        },
        "main": {
          "tagConfig": {
            "append": true,
            "hash": false,
            "publicPath": true,
            "tags": [
              {
                "attrs": {
                  "src": "jq.js",
                },
                "tag": "script",
              },
            ],
          },
        },
      },
    },
  ],
}
`;

exports[`plugin-html > should allow to modify plugin options by tools.htmlPlugin 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "meta": {},
        "minify": "auto",
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "auto",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
      "userOptions": {
        "inject": true,
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should allow to set favicon by html.favicon option 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": "src/favicon.ico",
        "filename": "index.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "favicon": "src/favicon.ico",
        "filename": "index.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should allow to set inject by html.inject option 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "body",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": "body",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should enable minify in production 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": {
          "collapseWhitespace": true,
          "keepClosingSlash": true,
          "minifyCSS": true,
          "minifyJS": {
            "format": {
              "ascii_only": true,
            },
            "mangle": {
              "safari10": true,
            },
          },
          "minifyURLs": true,
          "removeComments": false,
          "removeEmptyAttributes": true,
          "removeRedundantAttributes": true,
          "removeScriptTypeAttributes": true,
          "removeStyleLinkTypeAttributes": true,
          "useShortDoctype": true,
        },
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should register html plugin correctly 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should stop injecting <script> if inject is '() => false' 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": false,
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": false,
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should stop injecting <script> if inject is 'false' 1`] = `
{
  "entry": {
    "index": [
      "src/index.js",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "index",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "index",
        "excludeChunks": [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": false,
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "index",
        ],
        "entryName": "index",
        "filename": "index.html",
        "inject": false,
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "index": {},
      },
    },
  ],
}
`;

exports[`plugin-html > should support multi entry 1`] = `
{
  "entry": {
    "foo": [
      "src/foo.ts",
    ],
    "main": [
      "src/main.ts",
    ],
  },
  "plugins": [
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "main",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "main",
        "excludeChunks": [],
        "favicon": false,
        "filename": "main.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "main",
        ],
        "entryName": "main",
        "filename": "main.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlWebpackPlugin {
      "options": {
        "base": false,
        "cache": true,
        "chunks": [
          "foo",
        ],
        "chunksSortMode": "auto",
        "compile": true,
        "entryName": "foo",
        "excludeChunks": [],
        "favicon": false,
        "filename": "foo.html",
        "hash": false,
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "publicPath": "auto",
        "scriptLoading": "defer",
        "showErrors": true,
        "template": "<ROOT>/packages/core/static/template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Rsbuild App",
        "xhtml": false,
      },
      "userOptions": {
        "chunks": [
          "foo",
        ],
        "entryName": "foo",
        "filename": "foo.html",
        "inject": "head",
        "meta": {
          "charset": {
            "charset": "UTF-8",
          },
          "viewport": "width=device-width, initial-scale=1.0",
        },
        "minify": false,
        "scriptLoading": "defer",
        "template": "<ROOT>/packages/core/static/template.html",
        "templateParameters": [Function],
        "title": "Rsbuild App",
      },
      "version": 5,
    },
    HtmlBasicPlugin {
      "modifyTagsFn": undefined,
      "name": "HtmlBasicPlugin",
      "options": {
        "foo": {},
        "main": {},
      },
    },
  ],
}
`;
